habitat-lab/habitat/sims/habitat_simulator/kinematic_relationship_manager.py

```diff
@@ -0,0 +1,426 @@
+        self.ao_link_map = sutils.get_ao_link_id_map(self.sim)
+
+    def initialize_relationship_graph(self) -> None:
+        """
+        Scrape current scene contents to initialize the relationship graph.
+        #TODO: add configurable relationship sources (e.g. ontop vs. within vs. both)
+        """
+
+        self.relationship_graph = RelationshipGraph()
+        # do this once now instead of repeating for each ontop
+        self.sim.perform_discrete_collision_detection()
+
+        # NOTE: initialize implementation - use within set
+        # TODO: ontop would be good, but requires settling the sim, need a kinematic alternative
+        for obj_id in sutils.get_all_object_ids(self.sim).keys():
+            parent_obj = sutils.get_obj_from_id(
+                self.sim, obj_id, self.ao_link_map
+            )
+            assert parent_obj is not None, f"Object id {obj_id} is invalid."
+            # NOTE: because links are parented explicitly to their parent AO, we don't allow them to be children in the relationship manager.
+            # if parent_obj.object_id == obj_id:
+            #    #this is a ManagedObject, not a link
+            #    obj_within = sutils.within(self.sim, parent_obj)
+            #    for parent_id in obj_within:
+            #        self.relationship_graph.add_relation(parent_id, obj_id, "within")
+            # TODO: Using ontop here because it was easier to demo, this should change. Using more than one preposition makes cycles more common. Better to control this from the episode and task logic.
+            obj_ontop = sutils.ontop(
+                self.sim, obj_id, do_collision_detection=False
+            )
+            for child_id in obj_ontop:
+                if child_id == habitat_sim.stage_id:
+                    continue
+                child_obj = sutils.get_obj_from_id(
+                    self.sim, child_id, self.ao_link_map
+                )
+                if (
+                    child_id == child_obj.object_id
+                    and child_obj.motion_type
+                    != habitat_sim.physics.MotionType.STATIC
+                ):
+                    # this is a ManagedObject, not a link
+                    self.relationship_graph.add_relation(
+                        obj_id, child_id, "ontop"
+                    )
+
+        self.prev_snapshot = self.get_relations_snapshot()
+        self.prev_root_obj_state = self.get_root_parents_snapshot()
+
+    def _get_relations_recursive(
+        self, parent_id: int, wip_snapshot: Dict[int, Dict[int, mn.Matrix4]]
+    ) -> None:
+        """
+        Gather the relative transforms for the "work in progress" snapshot relations recursively to all children of the parent_id.
+
+        :param parent_id: The parent of the subtree on which to recurse.
+        :param wip_snapshot: The work-in-progress snapshot being constructed by this recursive process. default_dict(lambda: {})
+        """
```

Comment:
Is the snapshot still "wip"?

Reply:
This parameter is `wip_snapshot` because it is being constructed during recursion, so the snapshot is not complete yet. Not that the feature itself is "wip".